Combining stacks from an inventory no longer leaves bugged sprites

zkxs opened the pull request:
:cl:
bugfix: Combining stacks from an inventory no longer leaves bugged sprites
/:cl:

Fixes #27059

Tested this locally by tossing some stacks around and it seems like it works as intended.

## So, here's my understanding of what's happening currently

First, `/obj/item/stack/proc/transfer_to(obj/item/stack/S, var/tamount=null, var/type_verified)` is called.

This calls `/obj/item/stack/proc/use(var/used)`, which contains this snippet:
```
if (amount <= 0)
	qdel(src)
```

`qdel` eventually hooks into `/obj/item/Destroy()`, which contains this snippet:
```
var/obj/item/weapon/storage/storage = loc
if(istype(storage))
	storage.on_item_deletion(src)
```

`/obj/item/weapon/storage/proc/on_item_deletion(obj/item/W)` makes calls to the following:
```
/datum/storage_ui/default/on_pre_remove(var/mob/user, var/obj/item/W)
/datum/storage_ui/default/prepare_ui()
/datum/storage_ui/default/on_post_remove(var/mob/user)
```

- `on_pre_remove` removes the item from the screen
- `prepare_ui` iterates over `storage.contents`, rendering each item, _including the item which we are in the process of removing, and still exists in `contents`._ At this point, the (now empty) stack item get shoved right back onto `screen`, because the storage ui thinks it's still in the storage.
- finally, `on_post_remove` does nothing of interest.

After all this, the Destroy() continues to travel back up the `..()` chain until garbage collection eventually does whatever it does.

## How to fix it

~~Given my understanding of this bug, it seems like the simple fix is simply doing a `contents -= W` in `/obj/item/weapon/storage/proc/on_item_deletion(obj/item/W)`~~

Hot take: the secret sauce is that `on_pre_remove` needs to be called before `qdel` and `prepare_ui` needs to be called after.
- `on_pre_remove` needs the item to still exist so that it can be referenced and removed from the screen
- `prepare_ui` needs the item to be deleted, which apparently removes it from the storage's contents. I don't know _where_ this is happening, but it's definitely happening.

afterthought2 commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -182,6 +182,10 @@
+			if (istype(loc, /obj/item/weapon/storage))
+				var/obj/item/weapon/storage/S = loc
+				S.remove_from_storage(src)
 			qdel(src) //should be safe to qdel immediately since if someone is still using this stack it will persist for a little while longer
```
Should already be handled in here, through `on_item_deletion`.

zkxs replied:
Thanks for the feedback. I believe I have addressed this concern now.

afterthought2 commented on code/game/objects/items/weapons/storage/storage.dm:
```diff
@@ -244,6 +244,7 @@
 
 // Only do ui functions for now; the obj is responsible for anything else.
 /obj/item/weapon/storage/proc/on_item_deletion(obj/item/W)
+	contents -= W
```
This shouldn't be needed. If it is, it's not safe to do it like this; you could try and call `on_item_deletion` after the `..()` in `Destroy`.

zkxs replied:
I don't know why `contents -= W` isn't safe, but I certainly believe you. After much scratching of my head and thinking about your suggestion to call `on_item_deletion` after the `..()`, I think I've got it this time. I've updated the PR description to include my thought process.

zkxs committed:
Combining stacks from an inventory no longer leaves bugged sprites

:cl:
bugfix: Combining stacks from an inventory no longer leaves bugged sprites
/:cl:

Fixes #27059

Tested this locally by tossing some stacks around and it seems like it works as intended.
